### PR TITLE
Max states

### PR DESCRIPTION
**Pull request by martinhruska**
This PR implements logic of max state as described in PR #105.

### CODE REVIEW

File: include/mata/nfa.hh [hunk 1]
```diff
@@ -1071,7 +958,7 @@ bool are_equivalent(const Nfa& lhs, const Nfa& rhs, const Alphabet* alphabet,
  *
  * The current implementation of 'Mata::Nfa::Nfa' does not accept input alphabet. For this reason, an alphabet
  * has to be created from all transitions each time an operation on alphabet is called. When calling this function,
- * the alphabet has to be computed first.
+ * the alphabet contains to be computed first.
```
Comment (Adda0) [inline include/mata/nfa.hh]: Revert the change.

File: include/mata/nfa.hh [hunk 2]
```diff
@@ -1092,7 +979,7 @@ Nfa revert(const Nfa& aut);
 Nfa remove_epsilon(const Nfa& aut, Symbol epsilon = EPSILON);
 
 /// Test whether an automaton is deterministic, i.e., whether it has exactly
-/// one initial state and every state has at most one outgoing transition over
+/// one initial state and every state contains at most one outgoing transition over
```
Comment (Adda0) [inline include/mata/nfa.hh]: Revert the change.

File: bindings/python/libmata.pxd [hunk 3]
```diff
@@ -240,11 +240,11 @@ cdef extern from "mata/nfa.hh" namespace "Mata::Nfa":
...
-        void add_trans(CTrans) except +
-        void add_trans(State, Symbol, State) except +
-        void remove_trans(CTrans) except +
-        void remove_trans(State, Symbol, State) except +
-        bool has_trans(State, Symbol, State)
+        void add(CTrans) except +
+        void add(State, Symbol, State) except +
+        void remove(CTrans) except +
+        void remove(State, Symbol, State) except +
+        bool contains(State, Symbol, State)
```
Comment (Adda0) [inline bindings/python/libmata.pxd]: The Python binding has to be modified. This will not work as there would be multiple `add` operations, for example (one to add new state, other to add new transition etc.)

I would suggest reverting all changes to Python binding and @tfiedor or I will have to modify the Python binding manually.

File: bindings/python/libmata.pxd [hunk 4]
```diff
@@ -240,11 +240,11 @@ cdef extern from "mata/nfa.hh" namespace "Mata::Nfa":
         void unify_initial()
         void unify_final()
         COrdVector[Symbol] get_used_symbols()
-        void add_trans(CTrans) except +
-        void add_trans(State, Symbol, State) except +
-        void remove_trans(CTrans) except +
-        void remove_trans(State, Symbol, State) except +
-        bool has_trans(State, Symbol, State)
+        void add(CTrans) except +
```
Comment (tfiedor) [inline bindings/python/libmata.pxd]: Quite random rename. Is it wise? (EDIT: After reading rest, I understand the change, and have no issue)

File: src/nfa/nfa-complement.cc [hunk 5]
```diff
@@ -37,9 +37,9 @@ Nfa Mata::Nfa::Algorithms::complement_classical(
 	}
 
 	result = determinize(aut, subset_map);
-	State sink_state = result.states_number() + 1;
+	State sink_state = result.delta.post_size() + 1;
 	result.increase_size(sink_state+1);
-	assert(sink_state < result.states_number());
+	assert(sink_state < result.delta.post_size());
```
Comment (Adda0) [inline src/nfa/nfa-complement.cc:42]: Looking at how `Delta::post_size()` is used, I really am not a fan of the name for when you want to *get the number of automata states*. `post_size()` makes sense in the context of Delta itself, but not for this often used automata operation. I would not change the name inside Delta.

I would consider, however, adding an *alias* method on the whole NFA `NFA::states_number()` which would return `this.delta.post_size()`, simply to allow the user to type `state num` and find the method. Similarly, there could be `Nfa::transitions_number()` for transitions. But, this goes against our idea that we want to force the user to use the underlying classed for delta, initial and final states. Any ideas?

**Review by Adda0 — COMMENTED** [review]
I am OK with the PR. The suggestions are listed below.

File: include/mata/nfa.hh (hunk 2 above)
Comment (martinhruska) [reply include/mata/nfa.hh]: done

File: include/mata/nfa.hh (hunk 1 above)
Comment (martinhruska) [reply include/mata/nfa.hh]: done

File: bindings/python/libmata.pxd (hunk 3 above)
Comment (martinhruska) [reply bindings/python/libmata.pxd]: Okay, I reverted all changes in Python binding. Please do it manually.

File: src/nfa/nfa-complement.cc (hunk 5 above)
Comment (martinhruska) [reply src/nfa/nfa-complement.cc:42]: We should discuss this on thursday meeting.

File: src/nfa/nfa.cc [hunk 6]
```diff
@@ -406,15 +391,152 @@ void Nfa::remove_trans(State src, Symbol symb, State tgt) {
...
+    const Post& tl = post[src];
+    if (tl.empty()) {
+        return false;
+    }
+    auto symbol_transitions{ tl.find(Move{symb} ) };
+    if (symbol_transitions == tl.cend()) {
+        return false;
+    }
+
+    if (symbol_transitions->targets.find(tgt) == symbol_transitions->targets.end()) {
```
Comment (tfiedor) [inline src/nfa/nfa.cc]: Maybe could be rewritten to `return symbol_transitions->targets.find(tgt) != symbol_transitions->targets.end()`

File: src/nfa/nfa.cc [hunk 7]
```diff
@@ -509,7 +631,7 @@ bool Mata::Nfa::are_state_disjoint(const Nfa& lhs, const Nfa& rhs)
         if (haskey(lhs_states, rhs_st)) { return false; }
     }
 
-    for (size_t i = 0; i < lhs.delta.size(); i++) {
+    for (size_t i = 0; i < lhs.delta.post_size(); i++) {
```
Comment (tfiedor) [inline src/nfa/nfa.cc]: Maybe extract the call (was probably automatical batch rename and went unnoticed)

**Review by tfiedor — APPROVED** [review]
No major issues.

File: src/nfa/nfa.cc (hunk 6 above)
Comment (martinhruska) [reply src/nfa/nfa.cc]: yes, that would be better. Done.

File: src/nfa/nfa.cc (hunk 7 above)
Comment (martinhruska) [reply src/nfa/nfa.cc]: done

**Comment by martinhruska** [comment]
@tfiedor could you please fix the binding.

**Comment by tfiedor** [comment]
Yes, I will during today or tomorrow.

File: src/nfa/nfa-complement.cc (hunk 5 above)
Comment (kilohsakul) [reply src/nfa/nfa-complement.cc:42]: I also did not expect that it would be used everywhere. I thought we would use nfa.max_state(), or what the name if that is, everywhere, and use post_size() very seldom, essentially only for determining the nfa.max_state().

**Comment by tfiedor** [comment]
There is some segmentation fault in `make_complete`. Any ideas, what is wrong? I'm afraid it is probably not an issue of binding but of library. @martinhruska @Adda0 ?

File: src/nfa/nfa.cc [hunk 8]
```diff
@@ -330,15 +330,21 @@ void OnTheFlyAlphabet::add_symbols_from(const StringToSymbolMap& new_symbol_map)
...
+    return res;
+}
 
-void Nfa::add_trans(State state_from, Symbol symbol, State state_to) {
-    // TODO: Define own exception.
-    if (!is_state(state_from) || !is_state(state_to)) {
-        throw std::out_of_range(std::to_string(state_from) + " or " + std::to_string(state_to) + " is not a state.");
+void Delta::add(State state_from, Symbol symbol, State state_to)
+{
+    if (state_from >= post.size()) {
```
Comment (kilohsakul) [inline src/nfa/nfa.cc:343]: little thing, maybe using push_back instead of resize might be more robust wrt change of a data structure (like vector to deque)

File: src/nfa/nfa.cc [hunk 9]
```diff
@@ -542,7 +661,7 @@ void Mata::Nfa::make_complete(
         worklist.pop_front();
 
         std::set<Symbol> used_symbols;
-        if (!aut.has_no_transitions())
+        if (!aut.delta.has_no_transitions())
```
Comment (kilohsakul) [inline src/nfa/nfa.cc]: could it be delta.empty() ?

**Comment by Adda0** [comment]
> There is some segmentation fault in make_complete. Any ideas, what is wrong? I'm afraid it is probably not an issue of binding but of library.

If you could provide a reproducible example, ideally in Mata Catch2 format, that would be great. Otherwise, without any further information, no idea. It is probably connected to the recent changes, but who knows.

File: include/mata/nfa.hh [hunk 10]
```diff
@@ -645,7 +600,7 @@ public:
      */
     void unify_final();
 
-    bool is_state(const State &state_to_check) const { return state_to_check < delta.size(); }
+    bool is_state(const State &state_to_check) const { return state_to_check <= delta.max_state(); }
```
Comment (kilohsakul) [inline include/mata/nfa.hh]: Should we ask the automaton about his max_state rather than delta?

File: src/nfa/nfa.cc [hunk 11]
```diff
@@ -355,40 +361,19 @@ void Nfa::add_trans(State state_from, Symbol symbol, State state_to) {
...
-            symbol_transitions->insert(states_to);
-            return;
-        } else {
-            // Add transitions to a new Move struct with symbol yet unused on transitions from state_from.
-            const Move new_symbol_transitions {Move{symbol, states_to }};
-            state_transitions.insert(new_symbol_transitions);
-        }
+void Delta::remove(State src, Symbol symb, State tgt) {
+    if (src >= post.size()) {
+        throw std::runtime_error("State " + std::to_string(src) + " is not in Delta");
```
Comment (kilohsakul) [inline src/nfa/nfa.cc]: throw exception or do nothing?

File: src/nfa/nfa.cc [hunk 12]
```diff
@@ -406,15 +391,148 @@ void Nfa::remove_trans(State src, Symbol symb, State tgt) {
...
+    }
+    auto symbol_transitions{ tl.find(Move{symb} ) };
+    if (symbol_transitions == tl.cend()) {
+        return false;
+    }
+
+    return symbol_transitions->targets.find(tgt) != symbol_transitions->targets.end();
+}
+
+bool Delta::has_no_transitions() const
```
Comment (kilohsakul) [inline src/nfa/nfa.cc]: empty ?

**Comment by tfiedor** [comment]
> > There is some segmentation fault in make_complete. Any ideas, what is wrong? I'm afraid it is probably not an issue of binding but of library.
> 
> If you could provide a reproducible example, ideally in Mata Catch2 format, that would be great. Otherwise, without any further information, no idea. It is probably connected to the recent changes, but who knows.

Yep, I was planning on that. I pushed the test case, that causes segfault. So the library should be broken now.

**Comment by tfiedor** [comment]
@martinhruska @Adda0, i fixxed the segfault. There is a minor issue with binding. I will fix it soon.

File: include/mata/nfa.hh [hunk 13]
```diff
@@ -374,65 +381,49 @@ public:
...
-                renaming[i] = last_empty;
-                last_empty = i;
-            } else { // there was no empty space, last_empty is synchronized with i
-                renaming[i] = i; // nothing changed, no renaming needed
-                last_empty += 1;
-            }
-        }
+        post.resize(n);
+        if (post.size()-1 > max_state_)
+            max_state_ = post.size();
```
Comment (tfiedor) [inline include/mata/nfa.hh]: I'm debuggin some issues in Python binding, and I think, there might be an error. Shouldn't this be `max_state_ = post.size() - 1`? Because, if we resize the post to e.g. `12`, then the max state should be `11`, right? Since, we count from `0` @martinhruska can you check if my reasoning is correct? This would fix the issues in binding as well, I think.

Comment (martinhruska) [reply include/mata/nfa.hh]: yes, you are probably right, it should be `max_state_ =  post.size()-1` because counting from zero.

File: src/nfa/nfa-complement.cc (hunk 5 above)
Comment (martinhruska) [reply src/nfa/nfa-complement.cc:42]: Ok, i'm gonna check where can be `post_size` hidden a replaced by `max_state()`

File: src/nfa/nfa.cc (hunk 8 above)
Comment (martinhruska) [reply src/nfa/nfa.cc:343]: you mean applying push_back `state_from - post.size()` times. Because I think we don't have guaranteed that `state_from` is bigger just by one.

File: src/nfa/nfa.cc (hunk 9 above)
Comment (martinhruska) [reply src/nfa/nfa.cc]: `empty()` has different semantics. `has_no_transitions()` is appropriate here. Technically, empty means that there are no states in `post`, however in current implementation you can have states in post without any transitions. Therefore we need `has_no_transitions()`.

File: src/nfa/nfa.cc (hunk 12 above)
Comment (martinhruska) [reply src/nfa/nfa.cc]: `empty` is there already. It is different thing as I described above.

**Comment by tfiedor** [comment]
@martinhruska who will fix the bug? Will, you fix it together with potentially addressing Lukas's feedback? Or, you won't be addressing anything? If not, I could potentially fix the bug myself. Just so we have no conflicts. 

Currently, I'm unlocking the fixing of the binding (first the bug in the library should be fixed).

File: src/nfa/nfa.cc (hunk 11 above)
Comment (martinhruska) [reply src/nfa/nfa.cc]: I think it's matter of taste. I prefer throwing exception, it can help in debugging.

File: include/mata/nfa.hh (hunk 10 above)
Comment (martinhruska) [reply include/mata/nfa.hh]: this makes sense in the current implementation. Need to check whether it does not break anything.

**Comment by martinhruska** [comment]
I'm gonna fix it. So can I start commiting changes?

**Comment by tfiedor** [comment]
Yes, go for it. Ping me, when your done. If the binding will still be broken, I will fix it, if not, you can merge it.

File: include/mata/nfa.hh (hunk 10 above)
Comment (martinhruska) [reply include/mata/nfa.hh]: solved.

**Comment by martinhruska** [comment]
@kilohsakul please check my answers to your comments.

File: src/nfa/nfa-complement.cc (hunk 5 above)
Comment (martinhruska) [reply src/nfa/nfa-complement.cc:42]: Well, my IDE shows around 90 usages of `post_size`. I would rather make refactoring to using `max_state` in another PR.

**Comment by tfiedor** [comment]
> @kilohsakul please check my answers to your comments.

Are you done? Should I fix the binding?

**Comment by martinhruska** [comment]
So far, I'm done. However, they may be other changes based on discussion with Lukas.

**Comment by tfiedor** [comment]
@martinhruska, the binding is fixed. Also, note that I fixed one bug in the library (in `Nfa::clear()`, the `max_state_` should be cleared to 0 as well.

Also, I think there could be more bugs regarding the `max_state_` and e.g. calling `resize()`, maybe we could discuss it tomorrow.

**Comment by martinhruska** [comment]
TODO:

- [x] Remove empty from `delta`, and then rename `has_no_transitions` to `delta`
- [x] remove `resize`
- [x] do nothing when accessing state in delta which is not in delta, or return empty post
- [x] add issue on using `nfa` max state instead of `delta` transitions.

**Comment by martinhruska** [comment]
I consider this PR as finished if nobody protests.  I would like to merge it and fully concentrate on that CMake thing.

**Comment by martinhruska** [comment]
I changed `operator[]` semantics to perform resize and added some tests.

**Comment by martinhruska** [comment]
can i merge it? Any objections?

**Review by Adda0 — APPROVED** [review]
Looks good to me. As in, let's merge it now and see how it works.
